flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -299,7 +296,7 @@ public Task(
 		@Nonnull TaskMetricGroup metricGroup,
 		ResultPartitionConsumableNotifier resultPartitionConsumableNotifier,
 		PartitionProducerStateChecker partitionProducerStateChecker,
-		Executor executor) {
+		Executor executor) throws IOException {
```

Comment:
It seems that partition/gate setup causes the exception. Usually, constructors should be able always just construct objects without doing any IO and throwing exceptions. If partition/gate had a lifecycle `setup()` interface method, which would do the same as `NetworkEnviroment.setupPartition/Gate`, then the setup could stay in the same place in `Task.run` as originally, do IO and throw exceptions while running.

Reply:
Actually I also do not want to throw execution during constructing `Task`, so in the first version I introduced `Task#initializeInputOutput` for creating and setup partition/gate together. But after reference with partition/gate in task, I reverted the codes as now. I have some concerns for adding `setup` in the interface of partition/gate.

1. I ever thought the lifecycle of partition/gate should be managed inside `ShuffleService`, like `create`, `setup` and `close`, which seems not leaked and self-explanation for different `ShuffleService` implementation. But it would bring overhead for adding `setup` in `ShuffleService` interface.

2. If task could call `setup` of partition/gate directly, it still needs add the partition/gate interface method which might not necessary for other implementations. In addition we need pass more parameters for this `setup`, like `NetworkBufferPool`, `buffersPerGate` etc.

If we add the `setup` lifecycle for task instead of partition/gate, which is similar as my previous `Task#initializeInputOutput`, it might seem easy because we only have one `Task` implementation. In addition, I wish to pass `BufferPool` directly in the construction of partition/gate instead of separate `registerBufferPool` method in future.  What do you think?